### PR TITLE
add clustering coefficients, join, core_periphery_deg

### PR DESCRIPTION
**Pull request by CarloLucibello**

### CODE REVIEW

**Comment by jpfairbanks** [comment]
for the clustering coefficient: have you looked at using set intersection rather than checking each edge individual? Which one is faster?

**Comment by CarloLucibello** [comment]
Seems to be slower. With this version
```julia
function global_clustering(g::SimpleGraph)
    c = 0
    ntriangles = 0
    for v in 1:nv(g)
        neighs = neighbors(g, v)
        eset = Set([Edge(i,j) for i=neighs,j=neighs])
        c += 2*length(intersect(edges(g), eset))
        # for i=1:length(neighs)
        #     for j=1:length(neighs)
        #         i == j && continue
        #         if has_edge(g, neighs[i], neighs[j])
        #             c += 1
        #         end
        #     end
        # end
        k = degree(g, v)
        ntriangles += k*(k-1)
    end
    ntriangles == 0 && return 1.
    return c / ntriangles
end
```
we have
```julia
julia> @time global_clustering(CompleteGraph(100))
  0.271835 seconds (172.88 k allocations: 85.213 MB, 4.61% gc time)
1.0

julia> @time global_clustering(CompleteGraph(100))
  0.120286 seconds (7.46 k allocations: 77.989 MB, 5.25% gc time)
1.0
```

while previous version gives
```julia
julia> @time global_clustering(CompleteGraph(100))
  0.172214 seconds (131.26 k allocations: 6.382 MB)
1.0

julia> @time global_clustering(CompleteGraph(100))
  0.062930 seconds (1.46 k allocations: 816.625 KB)
1.0
```

Probably a factor 2 can be gained in both versions specializing for Graph and playing with the indexes, but for the time being I don't think it is worth the effort

**Comment by sbromberger** [comment]
This:

```
eset = Set([Edge(i,j) for i=neighs,j=neighs])
```
will be expensive.

**Comment by CarloLucibello** [comment]
> I don't know why has_edge() gets such a bad rap here. With edge sets it's LITERALLY O(1). Use it in good health.

I'm beginning to appreciate it :)

this is ready for merge

**Comment by sbromberger** [comment]
Thanks. We have some conflicts since your PRs are in the middle of a substantial move to v0.4. Would you mind resolving the conflicts (pull from master first) and recommitting? That will also give me enough time to finish tagging v0.4.0. :)

**Comment by CarloLucibello** [comment]
done! It was a bit messy though, I hopoe I didn't lost anything

File: test/runtests.jl [hunk 1]
```diff
@@ -42,6 +42,12 @@ a1 = Graph(adjmx1)
 a2 = DiGraph(adjmx2)
 
 tests = [
+    "operators",
+    "community/core-periphery",
+    "community/detection",
+    "community/modularity",
+    "community/clustering",
+    "randgraphs",
     "graphdigraph",
```
Comment (sbromberger) [inline test/runtests.jl]: Gah. Sorry - one last change. Can you please preserve the oder of runtests? That is, graphdigraph should be first, operators should be near the top, and community/ stuff should be towards the end. I know it's useful for testing pre-merge, but we should switch it back before committing.

**Comment by CarloLucibello** [comment]
modified according to james' suggestion in #200

**Comment by sbromberger** [comment]
Looks like we need more tests...

File: src/community/core-periphery.jl [hunk 2]
```diff
@@ -0,0 +1,21 @@
+"""
+A simple degree-based core-periphery detection algorithm (see [Lip](http://arxiv.org/abs/1102.5511)).
+Returns the vertex assignments (1 for core and 2 for periphery).
+"""
+function core_periphery_deg(g::Graph)
+    degs = degree(g)
+    π = sortperm(degs, rev=true)
```
Comment (sbromberger) [inline src/community/core-periphery.jl]: Can we not use "pi"?

**Comment by CarloLucibello** [comment]
should be ready now

**Comment by sbromberger** [comment]
Awesome work. Thanks, @CarloLucibello!

**Pull request merged**